applets/wncklet/wayland-backend.c

```diff
@@ -207,7 +209,27 @@ foreign_toplevel_handle_app_id (void *data,
 				struct zwlr_foreign_toplevel_handle_v1 *toplevel,
 				const char *app_id)
 {
-	/* ignore */
+	ToplevelTask *task = data;
+
+	gchar *app_id_lower = g_utf8_strdown (app_id, -1);
+	gchar *desktop_app_id = g_strdup_printf ("%s.desktop", app_id_lower);
```

Comment:
This is unfortunately not complete. In some cases, especially with X11 apps, the app ID is not the name of the .desktop file but the value of the `StartupWMClass` key inside it.

Reply:
While this may be true it is not "best practice" to define the wayland app_id, at least based on the comment in xdg-shell.xml [1]:
> As a best practice, it is suggested to select app
> ID's that match the basename of the application's .desktop file.
> For example, "org.freedesktop.FooViewer" where the .desktop file is
> "org.freedesktop.FooViewer.desktop".

I think this issue should instead be handled at the xwayland bridge so the app_id it generates based on WmClass points to a valid .desktop basename (if one exists).

Doing a lowercase of the app_id seems not best practice either though. I just assumed it would be because the app_id I got for pluma was "Pluma" but the .desktop file for pluma is actually called pluma.desktop.

In any case this is better than not showing any icons at all and can be changed later on.
The  only downside of not parsing *all* desktop files to match based on StartupWMClass is that some icons may show up with the theme's default broken image.

[1] https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/stable/xdg-shell/xdg-shell.xml#L654

Reply:
Yes, for native Wayland apps. There are however still X11-only apps that run using XWayland, and X11's `WM_CLASS` is of course a bit different from Wayland's `app_id`. I do however agree that only handling native Wayland apps' `app_id` properly is better than not handling icons at all on Wayland.

Reply:
It's been a year and a half, we can fix the rest of this later. Icons are now working with this